Dpp 1330 simplify payments due cbm

carlbm opened the pull request.

carlbm committed:
Added tests to exersise the datalock output duplicate removal

carlbm committed:
Added automatic loading of tests from excel, with the results in excel

carlbm committed:
Added first day of academic year to payment calculations - essential to get the correct price episode identifier for Math/English that are paid in an academic year without onprog payments

carlbm committed:
Added tests for overlapping datalocks

carlbm committed:
Added test for ACT1 maths/english payment with no onprog payment in the year

carlbm committed:
Added tests for ACT1 maths/english, and ACT1 maths/english with no on prog in year

carlbm committed:
Added extra equality comparer operators for datalock output and match set for payments

carlbm committed:
Merge remote-tracking branch 'origin/DPP-1330-simplify-payments-due' into DPP-1330-simplify-payments-due-cbm

Pauljgraham commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/DomainTests/MatchSetTests/GivenAMatchSet.cs:
```diff
@@ -1,13 +1,92 @@
+        }
+
+        [Theory, PaymentsDueAutoData]
+        public void ThenTwoDifferentObjectsWithIdenticalPropertiesWillBeEqual(
```
This looks like a DatalockOutput test

Pauljgraham commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/DomainTests/DatalockTests/GivenADatalockOutput.cs:
```diff
@@ -0,0 +1,112 @@
+
+            actual.Should().BeFalse();
+        }
+    }
```
Would it be worthwhile testing identical tests except where teh commitment version id is different and the other properites we do not compare

Pauljgraham commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/DomainTests/LearnerTests/GivenAnAct1LearnerChangesPrice.cs:
```diff
@@ -47,7 +47,7 @@ public void ThereArePaymentsForR01()
             var datalock = new IShouldBeInTheDatalockComponent();
             var datalockResult = datalock.ValidatePriceEpisodes(_commitments, _datalocks, _datalockValidationErrors,
-                _earnings.Take(1).ToList(), _mathsAndEnglishEarnings);
+                _earnings.Take(1).ToList(), _mathsAndEnglishEarnings, new DateTime(2017, 08, 01));
```
Not sure if this will be a problem when we move to another academic year (I haven't looked into the new code)

Pauljgraham commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Domain/DatalockOutput.cs:
```diff
@@ -10,7 +10,7 @@ public DatalockOutput(DatalockOutputEntity entity)
             Ukprn = entity.Ukprn;
             PriceEpisodeIdentifier = entity.PriceEpisodeIdentifier;
-            LearnRefNumber = entity.LearnRefNumber;
+            LearnRefNumber = entity.LearnRefNumber??string.Empty;
```
Why treat null as a an empty string?

Pauljgraham commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Domain/IShouldBeInTheDatalockComponent.cs:
```diff
@@ -174,7 +196,26 @@ private void MatchMathsAndEnglishToOnProg()
                 if (RawEarningsMathsOrEnglish.All(x => x.ApprenticeshipContractType == 2))
                 {
-                    MarkNonZeroTransactionTypesAsPayable(RawEarningsMathsOrEnglish);
+                    var firstDayOfNextAcademicYear = _firstDayOfAcademicYear.AddYears(1);
```
This code is quite hard to understand (but I know it's never been easy logic)

Pauljgraham commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Domain/MatchSetForPayments.cs:
```diff
@@ -84,5 +64,54 @@ public override int GetHashCode()
             }
         }
+
+        public static bool operator ==(MatchSetForPayments left, MatchSetForPayments right)
```
What is the purpose of MatchSetPayments? I thought we were just adding unit tests

Pauljgraham commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Domain/MatchSetForPayments.cs:
```diff
@@ -1,6 +1,8 @@
+namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.Domain
 {
-    public class MatchSetForPayments
+    public class MatchSetForPayments : IEquatable<MatchSetForPayments>
```
This is used as a GroupBy Key, would it be better to call it PaymentGroupKey

carlbm committed:
Added failing test for production scenario

carlbm committed:
Moved datalock validation and commitment matching to a seperate service

carlbm committed:
Renaming to better match intent/behaviour

carlbm committed:
Major refactor
Moved learner to a service called CalculatePaymentsDue, no longer has parameters passed into ctor, now passes in the method

carlbm committed:
Made DatalockOutput and PaymentGroup sealed
Added tests for IDetermineWhetherEarningsArePaid
Extended the setup code to handle maths/english

carlbm committed:
Fixed failing tests. Removed old sql files

carlbm committed:
Added test to get 100% coverage for the logic in IDecideWhetherToPayEarnings

carlbm committed:
Oops. Sql files were not old. Repaced them

carlbm committed:
Updated logger to the correct logger

carlbm committed:
Merge remote-tracking branch 'origin/DPP-1330-simplify-payments-due-learner-behaviour' into DPP-1330-simplify-payments-due-cbm

carlbm committed:
Merge remote-tracking branch 'origin/DPP-1330-simplify-payments-due' into DPP-1330-simplify-payments-due-cbm

carlbm committed:
Revereted change to refunds for levy payments

carlbm committed:
Revert "Fixed co-invested refund processor to refund previous months if not enough payments"

This reverts commit 4db774fcc31d66c8255f5170827c18b91b1c2678.

carlbm committed:
Added tests for rerunning the decider and payer methods

carlbm committed:
Small change to test. Doesn't make it pass, but it seems correct

carlbm committed:
Added more tests for DatalockOutput equality (or not testing account id, account version or commitment version for equality)

carlbm committed:
Merge

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.IntegrationTests/Infrastructure/GivenANonPayableEarningRepository.cs:
```diff
@@ -174,8 +174,8 @@ public new void Setup()
 
                 [Test]
                 public void ThenItSetsReason() =>
```
update the test name

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/DomainTests/DatalockTests/GivenADatalockOutput.cs:
```diff
@@ -0,0 +1,148 @@
+            actual.Should().BeFalse();
+        }
+    }
+}
```
its also possible to keep the test names to the given-when-then order, e.g.
`ThenTwoDifferentObjectsWithDifferentPropertiesWillNotBeEqual` 
can be changed to
`WhenEvaluatingEqualityOfTwoDifferentObjectsWithDifferentPropertiesThenWillNotBeEqual`
or
`WhenEvaluatingEqualityOfTwoDifferentObjectsWithDifferentPropertiesThenNotEqual`
and the `WhenEvaluatingEquality` part could be put into a nested test class to remove the repetition, 
`GivenADatalockOutput.WhenEvaluatingEquality.AndObjectsHaveDifferentPropertiesThenNotEqual`

carlbm replied:
Updated

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/DomainTests/DetermineWhichEarningsShouldBePaidTests/MathsAndEnglish/GivenADecider.cs:
```diff
@@ -0,0 +1,125 @@
+namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests.DomainTests.DetermineWhichEarningsShouldBePaidTests.MathsAndEnglish
+{
+    [TestFixture]
+    public class GivenADecider
```
what is a decider? sut here is IDetermineWhichEarningsShouldBePaid.

carlbm replied:
Updated class name and test names

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/DomainTests/LearnerTests/GivenAnAct1LearnerChangesPrice.cs:
```diff
@@ -1,110 +0,0 @@
-﻿using System;
```
if you rename a file in one commit and then make changes in a subsequent commit git will recognise that it is still the same file and keep history. if you rename and make (significant) changes git will assume it's a new file since it is so different from the previous version.

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/DomainTests/MatchSetTests/GivenAMatchSet.cs:
```diff
@@ -8,9 +8,124 @@ namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests.DomainTests.MatchS
     [TestFixture]
     public class GivenAMatchSet
     {
+        [Theory, PaymentsDueAutoData]
```
Why `Theory` and not `Test`? Not sure you are doing anything that warrants theory as you don't have any `Assume.That()` or `Datapoint`s.

carlbm replied:
Too much xunit... changed

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/GivenALearnerProcessParametersBuilder.cs:
```diff
@@ -45,7 +42,7 @@ public class GivenALearnerProcessParametersBuilder
             [Frozen] Mock<ICollectionPeriodRepository> collectionPeriodRepository,
             [Frozen] Mock<IRawEarningsRepository> mockRawEarningsRepository,
-            LearnerProcessParametersBuilder sut)
+            SortProviderDataIntoLearnerData sut)
```
test file name doesn't match class name.
class name sounds like a function name. I suggest `SortProviderDataIntoLearnerDataService`/`Provider`... other ideas?

carlbm replied:
Used ...Service

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/GivenALearnerProcessor.cs:
```diff
@@ -34,17 +29,11 @@ public class WhenCallingProcess
-                    .Setup(factory => factory.CreateDataLockComponent())
-                    .Returns(mockDataLockComponent.Object);
-
                 //mockDataLockComponent
```
remove the comments

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/GivenALearnerProcessor.cs:
```diff
@@ -67,18 +56,12 @@ public class WhenCallingProcess
-                    .Setup(factory => factory.CreateDataLockComponent())
-                    .Returns(mockDataLockComponent.Object);
-
                 //mockDataLockComponent
```
commented code is available in history, so kill it from the current version. same argument goes for any other commented code.

carlbm replied:
Removed file. Logic being tested in the Scenario tests

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/GivenAProviderProcessor.cs:
```diff
@@ -23,24 +23,24 @@ public class WhenCallingProcess
             public void ThenItGetsLearnerParametersForTheProvider(
             ProviderEntity provider,
-            [Frozen] Mock<ILearnerProcessParametersBuilder> mockParametersBuilder,
+            [Frozen] Mock<ISortProviderDataIntoLearnerData> mockParametersBuilder,
```
param name doesn't match type

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/PaymentsDueTask/WhenExecutingWithInvalidContext.cs:
```diff
@@ -5,7 +5,7 @@
 using NUnit.Framework;
 using SFA.DAS.Payments.DCFS.Context;
 using SFA.DAS.Payments.DCFS.Infrastructure.DependencyResolution;
-using SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests.Tools;
```
outil == tool in French.

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/ScenarioTesting/GivenALearnerWithAWithdrawnCommitment.cs:
```diff
@@ -1,13 +1,18 @@
 ﻿using System.Collections.Generic;
```
this file is an example where git has picked up the file name change correctly

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/ServiceTests/GivenACalculator/ThatIsUsedForMultipleCalculations.cs:
```diff
@@ -0,0 +1,32 @@
+using SFA.DAS.ProviderPayments.Calc.PaymentsDue.Services;
+using SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests.Utilities;
+
+namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests.ServiceTests.GivenACalculator
```
GivenAPaymentsDueCalculationService

carlbm replied:
In the folder name

carlbm replied:
In the test results (in R#) it's clear.

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests/Utilities/ReadOnlyClassManipulation/WriteToDatalockOutput.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.UnitTests.Utilities.ReadOnlyClassManipulation
+{
+    class WriteToDatalockOutput 
```
what's this doing?

carlbm replied:
SettingTheTransactionTypeFlag?

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Domain/ICanStoreCommitmentInformation.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.Domain
```
what's the difference between `ICanStoreCommitmentInformation` and `IHoldCommitmentInformation`.

carlbm replied:
One has getters, one has setters

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Services/PaymentsDueCalculationService.cs:
```diff
@@ -45,20 +40,21 @@ public List<RequiredPaymentEntity> CalculatePaymentsDue()
-                x.AccountId)).ToDictionary(x => x.Key, x => x.ToList());
+            var groupedPastPayments = pastPayments.GroupBy(x => new PaymentGroup
+                (
+                    x.StandardCode,
```
all of these assignment lines can be moved to an implicit cast operator overload. and tested in isolation. the same goes for lines 29-40.

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Services/Extensions/CommitmentExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.Services.Extensions
+{
+    public static class CommitmentExtensions
```
can be put onto the type and then tested.

carlbm replied:
It is tested. What else are you after?

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Infrastructure/Data/Repositories/DatalockRepository.cs:
```diff
@@ -18,7 +17,7 @@ public DatalockRepository(string connectionString) : base(connectionString)
-        public List<DatalockOutputEntity> GetDatalockOutputForProvider(long ukprn, DateTime firstDayOfNextAcademicYear)
+        public List<DatalockOutputEntity> GetDatalockOutputForProvider(long ukprn)
         {
             const string sql = @"
```
as the sql scripts are getting more involved, I suggest moving them to text files in the solution that are then referenced in code using a resource.

carlbm replied:
Probably need to talk about this one. Not sure what you're after. My view is that the SQL scripts are actually getting simpler

scottywakefield replied:
not so much for this pr tbh.

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Infrastructure/Data/Entities/RequiredPaymentEntity.cs:
```diff
@@ -1,11 +1,10 @@
+    public class RequiredPaymentEntity
     {
         public RequiredPaymentEntity()
         {}
```
is it necessary to have the constructor below?

carlbm replied:
Hmmmm. Mine doesn't have a constructor. Hoping that your code is not too far out of date ;)

scottywakefield commented on src/PeriodEnd/PaymentsDue/SFA.DAS.ProviderPayments.Calc.PaymentsDue/Domain/IDetermineWhichEarningsShouldBePaid.cs:
```diff
@@ -0,0 +1,362 @@
+namespace SFA.DAS.ProviderPayments.Calc.PaymentsDue.Domain
+{
+    // ReSharper disable once InconsistentNaming
+    public class IDetermineWhichEarningsShouldBePaid : IIDetermineWhichEarningsShouldBePaid
```
is this an interface or a class? maybe the interface should be IDetermineWhichEarningsShouldBePaid.cs and the class something else.

carlbm committed:
Merge remote-tracking branch 'origin/DPP-1330-simplify-payments-due' into DPP-1330-simplify-payments-due-cbm

carlbm committed:
Fixed up the formatting for AT specs where the tables were out
Added learnerprocessor to the scenario testing
Fixed ACT2 -> ACT1 test to better match the name of the test

carlbm committed:
Updated integration tests to include the renamed field and new field on NonPayableEarning

carlbm committed:
Updated names of the DatalockOutput tests

carlbm committed:
Updated name of IDetermineWhichEarningsShouldBePaid to DetermineWhichEarningsShouldBePaidService

carlbm committed:
Moved DetermineWhichEarningsShouldBePaidService to the services folder

carlbm committed:
Renamed test classes for DetermineWhichEarningsShouldBePaidService

carlbm committed:
Updated Theory to Test

carlbm committed:
Renamed SortProviderDataIntoLearnerData to SortProviderDataIntoLearnerDataService

carlbm committed:
Fixed provider processor tests and unignored them

carlbm committed:
Updated parameter name

carlbm committed:
Added constructor for PaymentGroup that accepts both funding due and requiredpaymententity.
